lib/new_relic/agent/log_event_attributes.rb

```diff
@@ -0,0 +1,123 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+module NewRelic
+  module Agent
+    class LogEventAttributes
+      MAX_ATTRIBUTE_COUNT = 240 # limit is 255, assume we send 15
+      ATTRIBUTE_KEY_CHARACTER_LIMIT = 255
+      ATTRIBUTE_VALUE_CHARACTER_LIMIT = 4094
+
+      attr_reader :custom_attributes, :already_warned_custom_attribute_count_limit
+
+      def initialize
+        @custom_attributes = {}
+        @already_warned_custom_attribute_count_limit = false
+      end
+
+      def add_custom_attributes(attributes)
+        return if already_warned_custom_attribute_count_limit
+
+        attributes.each do |key, value|
+          next if absent?(key) || absent?(value)
+
+          add_custom_attribute(key, value)
+        end
+      end
+
+      def reset!
```

Comment:
I have an idea to run by you... how about lazily setting the instance variables so that we don't have them taking up memory by default?

Here's what I'm thinking:

1) The `attr_reader` line gets completely removed

2) The `initialize` method gets completely removed

3) The `reset!` method changes to this:

```ruby
def reset!
  %i[custom_attributes already_warned_custom_attribute_count_limit].each do |attr|
    remove_instance_variable(attr) if instance_variable_defined?(attr)
  end
end
```

4) Define a new `custom_attributes` method:

```ruby
def custom_attributes
  @custom_attributes ||= {}
end
```

I'm thinking that those changes would be all that would be required with no other changes needed. The `@already_warned_custom_attribute_count_limit` instance var would be `nil` / undefined, which would evaluate as falsey, which we want by default.

Reply:
Great idea! Thanks for sharing. Incorporated in 3622f6a9b2a2ea8c74b3ec75ffac391fd303587d